list.c

```diff
@@ -0,0 +1,290 @@
+			else 
+			{
+				prev_iterator = (char*)((UINT*)iterator)[prev];
+				next_iterator = (char*)((UINT*)iterator)[next];
+				((UINT*)prev_iterator)[next] = (UINT)next_iterator;
+				((UINT*)next_iterator)[prev] = (UINT)prev_iterator;
+				free((void*)iterator);
+				iterator = next_iterator;
+			}
+		}
+		else
+		{
+			iterator = (char*)((UINT*)iterator)[next];
+		}
+	}
+}
+
+//value must be null string(end with '\0')
+void ListInsertNode(char ** list, const char * value)
+{
+	if (NULL != list)
+	{
+		char* node = NULL;
+
+		createNode(&node, value);
+
+		char* iterator = NULL;
+		char* next_iterator = NULL;
+
+		iterator = *list;
+
+		if (NULL == iterator)
+		{
+			*list = node;
+			return;
+		}
+
+		while (1)
+		{
+			next_iterator = (char*)((UINT*)iterator)[next];
+			if (NULL == next_iterator)
+			{
+				((UINT*)node)[prev] = (UINT)iterator; //add address to previous element
+				((UINT*)iterator)[next] = (UINT)node; //add address to next element
+				return;
+			}
+			iterator = next_iterator;
+		}
+	}
+	else printf("Something bad with parametrs:(");
+}
+	
+
+
+void PrintList(char** list)
+{
+	char* node = *list;
+	char* node_data = NULL;
+	if (NULL == node)
+	{
+		printf("List is empty!!!\n");
+	}
+	while (NULL != node)
+	{
+		node_data = (char*)(((UINT*)node)[data]);
+		printf("%s\n", node_data);
+		node = (char*)((UINT*)node)[next];
+	}
+}
+
+void ListRemoveDuplicates(char ** list)
+{
+	char* iterator = NULL;
+	char* next_iterator = NULL;
+	char* node_data_buffer = NULL;
+	iterator = *list;
+	while (NULL != ((UINT*)iterator)[next] && NULL != ((UINT*)*list)[next])
+	{
+		node_data_buffer = (char*)((UINT*)iterator)[data];
+		next_iterator = (char*)((UINT*)iterator)[next];
+		ListRemoveNode(&next_iterator, node_data_buffer);
+		iterator = next_iterator;
+	}
+}
+
+unsigned int ListSize(char ** list)
+{
+	char* iterator = *list;
+	UINT counter=0;
+	while (NULL != iterator)
+	{
+		++counter;
+		iterator = (char*)((UINT*)iterator)[next];
+	}
+	return counter;
+}
+
+int IndexOfFirstMatch(char ** list, const char * str)
+{
+	UINT counter = 0;
+	char* iterator = *list;
+
+	while (0 != strcmp((char*)((UINT*)iterator)[data], str))
+	{
+		if (NULL != (char*)((UINT*)iterator)[next])
+		{
+			++counter;
+			iterator = (char*)((UINT*)iterator)[next];
+		}
+		else
+		{
+			counter = -1;
+			break;  
+		}
+	}
+	
+	return counter;
+}
+
+void ListDestroy(char ** list)
+{
+	if (NULL != *list)
+	{
+		char* next_iterator = *list;
+		next_iterator = (char*)((UINT*)next_iterator)[next];
+		char* iterator = NULL;
+
+		while (NULL != next_iterator)
+		{
+			*list = (char*)((UINT*)next_iterator)[next];
+			free((void*)((UINT*)next_iterator)[data]);
+			free((void*)next_iterator);
+			next_iterator = (char*)((UINT*)*list)[next];
+		}
+
+		free((void*)((UINT*)*list)[data]);
+		free(*list);
+		*list = NULL;
+	}
+	else 
+	{
+		printf("List is already empty!");
+	}
+}
+
+void NodeSwap(char ** left, char ** right)//Just swap by value:(
+{
+	/*char* left_prev = NULL;
```

Comment:
Please remove commented code because for this purpose you have source control history